HiTRes Inversion Options

hanchawn opened the pull request:
Description:
- Add HiTRes options to hbmcmc set up to allow for creating mf timeseries using the HiTRes procedures
- Add edits to name to filter files imported for the date (avoid importing all matching files and then cropping data) and make some processes faster e.g. option to create timeseries in fp_sensitivity so timeseries_HiTRes doesn't need to be run both there and in footprints_data_merge
- Add overwrite and uncertainty options to flux.write

resolves issues #124 & #126

Type of Change:
[ ] Bug fix
[X] New feature
[ ] Code-breaking change

Checklist:
[ ] Code documentation has been updated if needed
[ ] If new tests are needed, they have been added
[X] All tests pass
[X] No merge conflicts
[X] /CHANGELOG.md has been updated if change is significant

hanchawn committed:
add options to allow HiTRes inversions and add species_footprints option


Former-commit-id: 524a305f1ef056fe4c3cf35e3448cc9bbc2d83dd

hanchawn committed:
Merge branch 'develop' into apo_inversion


Former-commit-id: 5fb9af139f2ccbf4456310bec3cfc1c394f5ca7d

hanchawn committed:
add HiTRes option which keeps all data until after timeseries is made and then filters out nans


Former-commit-id: f8c10d14bd19cdc7d27df1e8dba8025fa30a690b

hanchawn committed:
add HiTRes option which imports fluxes using the flux_for_HiTRes function


Former-commit-id: 4bcea91defcb8ee22b8751248a6edce5397d0e34

hanchawn committed:
update docstring for timeseries_HiTRes


Former-commit-id: 631950cf3c895f09883a0cf53a92412ea8cc1c24

hanchawn committed:
update footprints function for finding new HiTRes footprints and update docstrings


Former-commit-id: f3c90bd82ee15c6b49471ff70db24c2560d15314

hanchawn committed:
add option to flux_for_HiTRes for multiple sources


Former-commit-id: c6643f80a149c87cc7a7b5d91737e9f1920db65e

hanchawn committed:
fix bug with getting hitres fluxes for emds


Former-commit-id: ed627fbb17262fb3ec531e008f76333a3d0fb18e

hanchawn committed:
add chunking option to boundary_conditions function


Former-commit-id: c3dcd444f43274c95e373913c5b52a6b3b6f69b1

hanchawn committed:
add mol mass for apo


Former-commit-id: eb3002b5b40191792e5bb69b4528a09b81876e85

hanchawn committed:
add try except to reindexing in timeseries_HiTRes


Former-commit-id: 16d65e0c15dfb3e85482d62cb2f0a0b281d2e794

hanchawn committed:
check dates in filenames before importing bcs, fps, and fluxes to avoidusing too much memory


Former-commit-id: 364589fc2e9f03760c946b88151db4ac80da2ba1

Commit:
change building of timeseries in name.timeseries_HiTRes to avoid Implementation Error


Former-commit-id: 74f4a969d8f36b18a304d0fe2cdb09cef5bcc47a

hanchawn committed:
update import statements


Former-commit-id: c4ce8a647bb261234a61f53b21f2e6e743eceed8

hanchawn committed:
allow use of add_bc without footprints_data_merge


Former-commit-id: 55f5e1f86884a72fd98fff75615faed5b4566837

hanchawn committed:
add calc_timeseries option to fp_sensitivity so that we can avoid using timeseries_HiTRes twice if both are needed


Former-commit-id: 20239b5ed277cd9340185d03e2e207a6b5b48fad

hanchawn committed:
tidy up fp_sensitivity


Former-commit-id: 478b28e1bf671d2a69589cd377d9b895474be9c3

hanchawn committed:
avoid running timeseries_HiTRes twice, in footprints_data_merge and bc_sensitivity


Former-commit-id: c3dbde41bc5b9a8da6f6db8e863e125041dfae47

hanchawn committed:
use f strings in print statements


Former-commit-id: 54231427e57e1dd2dbe4b3daf2072fbe5e50dc9f

hanchawn committed:
add conversion for distance between lat and lon points in degrees to metres


Former-commit-id: 6bca8d9bd83937e02b35263d306704e1db242622

hanchawn committed:
edit latlon conversion


Former-commit-id: c82380d405b3a539b02733ca7a15c65fcb8debb4

hanchawn committed:
fix combine_diff_resolution to be able to use it with single value datasets


Former-commit-id: d68bd5630b911b08abb2d8b16d148b7f5deb7e9c

hanchawn committed:
add option to add a flux uncertainty array to flux.write


Former-commit-id: 034688e7f621e6bf95d9b2d2b704396c1dc1c1e4

hanchawn committed:
reduce lines in flux.write


Former-commit-id: 70b593cd09f815665b938642981ceb816fb1fc9b

hanchawn committed:
reduce lines in flux.write


Former-commit-id: 15bbfd42ebc93425773a82d2238fad7e92d42c92

hanchawn committed:
add chunk option to flux_for_HiTRes


Former-commit-id: 0ba832bbdd493e487a3773632777a8fea7ef03c8

hanchawn committed:
fix order that ts and mf map are returned by HiTRes timeseries in fp_sensitivity


Former-commit-id: 6135421929a10633b3dc1bf05fe74c6ab7827226

hanchawn committed:
change default chunks for fpdm to avoid chunking if not specified


Former-commit-id: 5331d20abcf7836c9f3030cb6ea78bbeea122f01

hanchawn committed:
add units for apo


Former-commit-id: d5616288860496c459aa1c5ba2c58b2f18239144

hanchawn committed:
change apo units to work with fpdm


Former-commit-id: e8b6316aed360a4aa5068f61d1d7fbee1baa8996

hanchawn committed:
fix string bug when checking dates in filenames to import fluxes


Former-commit-id: fb52211021a17dddbfcf5ffa5d46a4fcbc17d67d

hanchawn committed:
fix typo in cut_to_edge func in timevarying_BC.py


Former-commit-id: 67d0e942cbcc06c0420c4d2b8ad547628b6ad283

hanchawn committed:
correct typo in timevarying_BC cut_to_edge, and add a catch if overwriting an existing file


Former-commit-id: 9aeb82b6c64b2eeaffd0331d2e1b9d6882281be4

hanchawn committed:
fix catch for overwriting files in timevarying_BC


Former-commit-id: 1af45288a5a04a18fa43e47426d02c8c0729bf32

hanchawn committed:
add optional adjustment to timevarying_BCs for cases where there is an initital value which needs adding on


Former-commit-id: e429f6cad589d0d9eb5ef6d246d83e7de2bc398e

hanchawn committed:
Merge branch 'develop' into apo_inversion


Former-commit-id: 927042c5ae2a8490d846487180380360a2e9a9ea

hanchawn committed:
fix bug in lat values when converting degrees to m


Former-commit-id: 637f0b1089156ec98368764794d69ded61449fa3

hanchawn committed:
filter bc files by date needed
when running bc_sensitivity, if a start and end date are given then filter the files to match the dates. Also reformat the IOErrors to prevent big spaces in the printed text.


Former-commit-id: ada5b26580268ddb03b8ca9fb7e2ebfaa68ef513

hanchawn committed:
add start and end date when loading bc_sensitivity


Former-commit-id: f17ade2c14ef8b34e9e4c64c87b4fd90c783525a

hanchawn committed:
add function to name to filter files by date
:
- filter files to those that are within the start and end date range (if there are any) to avoid loading unneeded data
- fix bug with species_footprint to search for a generic footprint if species_footprint = None


Former-commit-id: d9d10fef2c418e480bb8f72699eac5bcd7c29152

hanchawn committed:
fix bc warning message
remove big space in warning message when bcs are not available for specified time period


Former-commit-id: a3b97408382bb897bdd53734f977e422e73a0fff

hanchawn committed:
print species_footprint if not None


Former-commit-id: e7900311be53a93d4c438b0f5433e237da2ad612

hanchawn committed:
add if for filtering basis bc files
 if there isnt a start and end date given then dont filter files


Former-commit-id: fdea8ff34c9330dc27fcfb36d313c4bb39eefef2

hanchawn committed:
allow for fpdm without fp var
- as some old HiTRes footprints have fp_HiTRes but not fp, allow for footprints_data_merge to work without fp


Former-commit-id: baf80ec46ec8e451a34a82203c57541a962edfd0

hanchawn committed:
take out bc sensitivity start and end dates
- filtering bc_sensitivty was breaking the code for an unknown reason, taking it out doesnt increase memory use that much so taking it out


Former-commit-id: 40e0f2f739b326903b6c10983c00dd27514a9d3d

hanchawn committed:
change file filtering

change the date that files are filtered by as it was importing then next years file when not needed, also add a HiTRes option to name.flux to get the correct time range


Former-commit-id: 0b2b72c07adc4a706f08d8feb3b1ed8dbc15c9f8

hanchawn committed:
fix file filtering


Former-commit-id: 574be44e5bf5bdd3232e8667b7a35a7a83528c03

hanchawn committed:
change the species entered for flux_for_HiTRes when creating emds array in inversionpymc3


Former-commit-id: c6c68c36a22c28a76793d13a441b8851d2f529b3

hanchawn committed:
Merge branch 'develop' into apo_inversion


Former-commit-id: 58dfa1f365110cf2397257a3c3f0a2dfd047f097

hanchawn committed:
take HiTRes option out of post processing
dont think HiTRres fluxes are needed for post processing - need to run a inversion to check that this is fine


Former-commit-id: 0f85b6b2279ed2810942ddde5a4c0a9792275754

hanchawn committed:
take out HiTres option for post-processing
post processing doesnt need any specific HiTRes options so take this back out

hanchawn committed:
Merge branch 'develop' into apo_inversion

hanchawn committed:
Search for generic fps if species fp cannot be found
  when the footprint for a specific species cannot be found then search for a footprint for generic species eg footprints for ch4 do not have ch4 in the filename

hanchawn committed:
take pixel-latlon conversion out
   haversine conversion already exists in a different file so take out repetition

hanchawn committed:
add overwrite option to flux.write
  add overwrite=False option to avoid asking user if they want to overwrite a file if it already exists

hanchawn committed:
print warning if overwriting file
   if overwrite=True and file is to be overwritten, tell user it is being overwritten

hanchawn committed:
take out HiTRes comments from hbmcmc processing

hanchawn committed:
use f-strings in flux.write

hanchawn committed:
add hbmcmc HiTRes option to changelog

hanchawn committed:
resolve merge conflict

hanchawn committed:
merge develop

hanchawn committed:
check for fp in site_ds when adding units

hanchawn committed:
convert high time res emissions_name
  if HiTRes=True in footprints_data_merge and emissions_name[source] is a string, convert to a dict for calculating the HiTRes timeseries

hanchawn committed:
flexible climatology date
change the dates for climatology in name.flux to allow for climatologies with a higher tim res than monthly

hanchawn committed:
add units option
   add option for units other than mol/m2/s

rt17603 commented:
I've had a quick look over this (will dig into more as well) but from a first glance it would be useful to have even a rudimentary test added for this functionality just to make sure everything works together ok.

hanchawn committed:
dont change start date for HiTRes
    in name.flux line 439, dont change the start date to start of year if getting HiTRes fluxes

hanchawn commented:
> 

In terms of the inversion, the HiTRes specific processes are only used in the set up i.e. where it calls footprints_data_merge, fp_sensitivity, bc_sensitivity - although I agree that the test coverage for the HiTRes part of name.py is not great
I thought that the existing hbmcmc tests would cover the rest as it's not HiTRes specific
The current tests pass for the changes I made to name.py, including a test for name.timeseries_HiTRes (but it doesn't test the other HiTRes parts of name.py)

hanchawn committed:
adjust flux_for_HiTRes date range
adjust the start datetime for name.flux_for_HiTRes to only import the required hours back, rather than a whole extra month

hanchawn committed:
add HiTRes tests
add tests for importing HiTRes footprints, and for name.flux_for_HiTRes

hanchawn committed:
import all data needed for HiTRes
initally import data from year before start if start month=1 and HiTRes=True, and then slice to required time range

hanchawn committed:
add H_back to footprints_data_merge
allow for flexible options of H_backwhen importing fluxes

rt17603 commented:
> > 
> 
> In terms of the inversion, the HiTRes specific processes are only used in the set up i.e. where it calls footprints_data_merge & fp_sensitivity - although I agree that the test coverage for the HiTRes part of name.py is not great I thought that the existing hbmcmc tests would cover the rest as it's not HiTRes specific The current tests pass for the changes I made to name.py, including a test for name.timeseries_HiTRes (but it doesn't test the other HiTRes parts of name.py)

Ok, but given that this PR is to integrate the HiTRes options into the hbmcmc code it would be useful to have even a very simple example which demonstrates that integration works. Even if we're happy each individual part works it's nice to know that the steps can and do work together.

hanchawn committed:
Merge branch 'develop' into apo_inversion

hanchawn committed:
avoid 2 asserts in 1 test
split HiTRes tests out where there were multiple asserts in 1 test to ensure all sserts are run if any fail

hanchawn committed:
add hitres ts test
   test that using HiTRes fp with constant flux gives the same as integrated fp with constant flux

hanchawn committed:
separate HiTRes ts setup
   this should allow for easier testing

hanchawn committed:
use diff to find timestep

hanchawn committed:
fix problem with getting Dec climatology

hanchawn committed:
use dask for flux_HiTRes in fpdm

hanchawn committed:
resolve merge conflicts

hanchawn committed:
remove unused module imports

hanchawn commented:
> 

I haven't added this test as I'm not sure how to write a test to run an inversion without it being very inefficient - from what I could see of the current hbmcmc tests they mostly use a toy model which doesn't use fpdm inputs (I could be wrong though).

It would be good to have this merged in so that we could add a version number to the APO paper, I can add an issue that some version of the test should be added later.

There are 4 tests failing but these are also failing in develop

rt17603 commented on acrg/BC/timevarying_BC.py:
```diff
@@ -32,15 +32,23 @@
     
 class BoundaryConditions:
     def __init__(self, vmr_var, gph_height_var, dataset=None, filename=None,
-                 file_dir=None, time_coord='time', height_coord='height', species=None,
```
Was there a reason the `height_coord` input was removed here? Is this no longer used?

hanchawn replied:
It looks like it was the same as gph_height_var, I've taken out the reamining refs to it

rt17603 commented on acrg/BC/timevarying_BC.py:
```diff
@@ -205,16 +215,16 @@ def cut_to_edge(self, fp_directory=None, fp_height_coord='height', convert_units
-        if self.dataset.coords['lon'].values[lon_e] < max(self.fp_lon) and lon_e != (len(dataset.coords['lon'].values)-1):
-            lat_s += 1
+        if self.dataset.coords['lon'].values[lon_e] < max(self.fp_lon) and lon_e != (len(self.dataset.coords['lon'].values)-1):
+            lon_e += 1
```
Were these bug fixes that were put in?

hanchawn replied:
Yes, the s was there where it should have been e

hanchawn committed:
remove height_coord

rt17603 commented on acrg/name/name.py:
```diff
@@ -288,38 +302,31 @@ def footprints(sitecode_or_filename, met_model = None, fp_directory = None,
-                       to integrated and HiTRes footprints \
-                       {integrated:path1, HiTRes:path2}")
-                return None
+    fp_directory = join(data_path, 'LPDM/fp_NAME/') if fp_directory is None else fp_directory
```
Based on previous code - is it no longer expected that `fp_directory` will be a dictionary for 'HiTRes' option?

hanchawn replied:
I've never used a dictionary for it as the HiTRes footprints are stored in the same area as the integrated ones now, do you think it would be better to change the code to make that an option again, or just take out reference to is being a dictionary?

hanchawn replied:
A lot of the code would need changing to use a dictionary for it, and I don't think you'd often have a reason for importing both so I think it would be best just to change the documentation
(The way the footprints are processed now means that the HiTRes code now only needs the HiTRes ones)

rt17603 replied:
I think the question is around backwards compatibility - thinking about it I think the dictionary option was because we used to have to use separate integrated and HiTRes files. If it's not too much work if may be good to allow it to be included as an option? Otherwise, maybe you could just add a comment to say this mode is no longer supported.

rt17603 replied:
Yeah ok I think that's fine - seems like a thing to add to the [change log](https://github.com/ACRG-Bristol/acrg/blob/develop/CHANGELOG.md) just so we know when the change happened.

hanchawn replied:
Sounds good

hanchawn committed:
rmeove ref to fp_directory as dict

hanchawn committed:
add to changelog

rt17603 commented:
I've reviewed this as much as I can given the time constraints and I think this looks really good overall.

You're correct that there are currently failing tests in develop anyway so these aren't being influenced by your changes. I think you're right it's best to get this in but as you say if an issue could be added to say that this needs to be tested that would be very helpful. At the very least were there some examples of how this can be run with example inputs etc.?

As a group we also should really think about getting this integrated into the `openghg-inversions` repository but that's a future problem and better to get this integrated for now.

hanchawn commented:
> 

Thank you Rachel, I appreciate it's a big PR so you'e done a thorough job in the circumstances!

I never got to the point of running a HiTRes inversion so I don't have example runs of that, but my APO models rely on some of the other changes to run.

Eric has added the HiTRes option to the openghg-inversions code, and is going to make any other changes that need to go in for that.
